// xiome/x/toolbox/snapstate/parts/tracking-mechanics.js
import { debounce } from "../../debounce/debounce.js";
import { debounceDelay } from "./constants.js";
import { SnapstateCircularError } from "./errors.js";
export function trackingMechanics() {
    let activeWatch;
    const sessions = new Map();
    function unsubscribe(identifier) {
        for (const [, record] of sessions)
            record.delete(identifier);
    }
    let waiter = Promise.resolve();
    return {
        get wait() { return waiter; },
        track(state, observer, reaction) {
            const identifier = Symbol();
            activeWatch = { identifier, session: { observer, reaction } };
            observer(state);
            activeWatch = undefined;
            return () => unsubscribe(identifier);
        },
        avoidCircular(key) {
            if (activeWatch)
                throw new SnapstateCircularError(`cannot set any properties within a reaction, not even "${key}"`);
        },
        triggerReactions: (() => {
            const reactionsQueue = new Set();
            const fireReactionsQueue = debounce(debounceDelay, (state) => {
                for (const key of reactionsQueue) {
                    for (const [, { observer, reaction }] of sessions.get(key) ?? []) {
                        if (reaction)
                            reaction(observer(state));
                        else
                            observer(state);
                    }
                }
                reactionsQueue.clear();
            });
            return (state, key) => {
                reactionsQueue.add(key);
                const promise = fireReactionsQueue(state);
                waiter = waiter.then(() => promise);
            };
        })(),
        reactionRegistration(key) {
            if (activeWatch) {
                let record = sessions.get(key);
                if (!record) {
                    record = new Map();
                    sessions.set(key, record);
                }
                record.set(activeWatch.identifier, activeWatch.session);
            }
        },
    };
}
