// xiome/x/toolbox/happystate/happy-combine.js
/*
 * happyCombine is tricky to use because typescript types are tricky.
 * example:
 *     const happy = happyCombine(happy1)(happy2)(happy3).combine()
 * do take note:
 *  - you must end the curried chain with `.combine()`
 *
 *
 */
export function happyCombine(happy1) {
    return function (happy2) {
        const getHappy = () => ({
            actions: { ...happy1.actions, ...happy2.actions },
            getState: () => ({ ...happy1.getState(), ...happy2.getState() }),
            clearStateListeners: () => {
                happy1.clearStateListeners();
                happy2.clearStateListeners();
            },
            onStateChange: (listener) => {
                const subscriptions = [
                    happy1.onStateChange(listener),
                    happy2.onStateChange(listener),
                ];
                return () => subscriptions.forEach(unsubscribe => unsubscribe());
            },
        });
        function more(happy3) {
            return happyCombine(getHappy())(happy3);
        }
        more.combine = getHappy;
        return more;
    };
}
