// xiome/x/features/videos/api/services/content-service.js
import * as renraku from "renraku";
import { getCatalog } from "./routines/get-catalog.js";
import { videoPrivileges } from "../video-privileges.js";
import { getAllViews } from "./routines/get-all-views.js";
import { getVideoViews } from "./routines/get-video-views.js";
import { concurrent } from "../../../../toolbox/concurrent.js";
import { DamnId } from "../../../../toolbox/damnedb/damn-id.js";
import { find } from "../../../../toolbox/dbby/dbby-helpers.js";
import { getDacastEmbed } from "./routines/get-dacast-embed.js";
import { getDacastApiKey } from "./routines/get-dacast-api-key.js";
import { getAllPrivileges } from "./routines/get-all-privileges.js";
import { getDacastContent } from "./routines/get-dacast-content.js";
import { setViewPermissions } from "./routines/set-view-permissions.js";
import { ingestDacastContent } from "./routines/ingest-dacast-content.js";
import { makePermissionsEngine } from "../../../../assembly/backend/permissions/permissions-engine.js";
import { makePrivilegeChecker } from "../../../auth/aspects/permissions/tools/make-privilege-checker.js";
export const makeContentService = ({ config, dacastSdk, videoTables: rawVideoTables, basePolicy, }) => renraku.service()
    .policy(async (meta, headers) => {
    const auth = await basePolicy(meta, headers);
    const appId = DamnId.fromString(auth.access.appId);
    const checker = makePrivilegeChecker(auth.access.permit, videoPrivileges);
    const engine = makePermissionsEngine({
        isPlatform: auth.access.appId === config.platform.appDetails.appId,
        permissionsTables: auth.authTables.permissions,
    });
    return {
        ...auth,
        engine,
        checker,
        videoTables: rawVideoTables.namespaceForApp(appId),
    };
})
    .expose(({ access, authTables, videoTables, checker, engine }) => ({
    async fetchModerationData() {
        checker.requirePrivilege("moderate videos");
        return concurrent({
            views: getAllViews({ videoTables }),
            privileges: getAllPrivileges({
                access,
                permissionsTables: authTables.permissions,
                platformAppId: config.platform.appDetails.appId,
            }),
            catalog: getDacastApiKey(videoTables)
                .then(async (apiKey) => apiKey
                ? getCatalog({ dacast: dacastSdk.getClient(apiKey) })
                : []),
        });
    },
    async writeView({ label, privileges, reference, }) {
        checker.requirePrivilege("moderate videos");
        await videoTables.viewDacast.update({
            ...find({ label }),
            whole: {
                label,
                type: reference.type,
                dacastId: reference.id,
            },
        });
        await setViewPermissions({
            label,
            engine,
            privileges,
            videoTables,
        });
    },
    async deleteView({ label }) {
        checker.requirePrivilege("moderate videos");
        await videoTables.viewDacast.delete(find({ label }));
        await videoTables.viewPrivileges.delete(find({ label }));
    },
    async getShows({ labels }) {
        const apiKey = await getDacastApiKey(videoTables);
        if (!apiKey)
            return [];
        const views = await getVideoViews({
            labels,
            checker,
            videoTables,
            userPrivileges: access.permit.privileges,
        });
        const dacast = dacastSdk.getClient(apiKey);
        return Promise.all(views.map(async (view, index) => {
            const label = labels[index];
            if (!view)
                return { label, details: undefined };
            const [data, embed] = await Promise.all([
                getDacastContent({ dacast, reference: view }),
                getDacastEmbed({ dacast, reference: view }),
            ]);
            const show = {
                label,
                details: data
                    ? {
                        ...ingestDacastContent({
                            type: view.type,
                            data,
                        }),
                        embed: embed?.code,
                    }
                    : null,
            };
            return show;
        }));
    },
}));
