// xiome/x/features/notes/api/notes-deposit-box.js
import { DamnId } from "../../../toolbox/damnedb/damn-id.js";
export function makeNotesDepositBox({ rando, notesTables }) {
    async function sendNotes(drafts) {
        const notes = drafts.map(draft => {
            const noteId = rando.randomId();
            return {
                noteBase: {
                    ...draft,
                    noteId,
                    to: DamnId.fromString(draft.to),
                    from: draft.from === undefined
                        ? undefined
                        : DamnId.fromString(draft.from),
                    old: false,
                    time: Date.now(),
                },
                noteDetails: draft.details,
            };
        });
        await notesTables.notes.create(...notes.map(n => n.noteBase));
        return notes.map(({ noteBase: { noteId } }) => ({
            noteId: noteId.toString()
        }));
    }
    async function sendNote(draft) {
        const [sent] = await sendNotes([draft]);
        return sent;
    }
    return {
        sendNotes,
        sendNote,
    };
}
