// xiome/x/features/xio-components/op/xio-op.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import styles from "./xio-op.css.js";
import { ops } from "../../../framework/ops.js";
import svgWarning from "../../../framework/icons/warning.svg.js";
import svgSpinner from "../../../framework/icons/spinner.svg.js";
import { AutowatcherComponent, html, mixinStyles, property } from "../../../framework/component.js";
let XioOp = class XioOp extends AutowatcherComponent {
    constructor() {
        super(...arguments);
        this.mode = "none";
        // @property({type: Object})
        // op: Op<xPayload> = ops.none()
        this._op = ops.none();
        this.errorIcon = svgWarning;
        this.loadingIcon = svgSpinner;
        this["loading-message"] = "loading...";
        this["error-message"] = "error";
    }
    get op() {
        return this._op;
    }
    set op(op) {
        const old = this._op;
        this._op = op;
        this.mode = ops.mode(op);
        this.requestUpdate("op", old);
    }
    render() {
        return ops.select(this.op, {
            none: () => html `
				<slot name=none></slot>
			`,
            loading: () => html `
				<slot name=loading>
					${this.loadingIcon}
					<span>${this["loading-message"]}</span>
				</slot>
			`,
            error: reason => html `
				<slot name=error>
					${this.errorIcon}
					<span>${reason ?? this["fallbackErrorMessage"]}</span>
				</slot>
			`,
            ready: () => html `
				<slot></slot>
			`,
        });
    }
};
__decorate([
    property({ type: String, reflect: true })
], XioOp.prototype, "mode", void 0);
__decorate([
    property({ type: Object })
], XioOp.prototype, "op", null);
__decorate([
    property()
], XioOp.prototype, "errorIcon", void 0);
__decorate([
    property()
], XioOp.prototype, "loadingIcon", void 0);
__decorate([
    property({ type: String })
], XioOp.prototype, "loading-message", void 0);
__decorate([
    property({ type: String })
], XioOp.prototype, "error-message", void 0);
XioOp = __decorate([
    mixinStyles(styles)
], XioOp);
export { XioOp };
