// xiome/x/framework/component/mixins/mixin-snapstate.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
export function mixinSnapstateTracking(...tracks) {
    return function (Base) {
        var _instances, _tracks, _unsubscribes, _subscribe, _sub, _unsub, _a;
        return _a = class extends Base {
                constructor() {
                    super(...arguments);
                    _instances.add(this);
                    _tracks.set(this, [...tracks]);
                    _unsubscribes.set(this, []);
                    _subscribe.set(this, (track) => {
                        return track(() => { this.render(); }, () => { this.requestUpdate(); });
                    });
                }
                render() {
                    return super.render && super.render();
                }
                attachTracking(...newTracks) {
                    __classPrivateFieldGet(this, _tracks, "f").push(...newTracks);
                    __classPrivateFieldGet(this, _instances, "m", _sub).call(this, newTracks);
                }
                connectedCallback() {
                    super.connectedCallback();
                    __classPrivateFieldGet(this, _instances, "m", _sub).call(this, __classPrivateFieldGet(this, _tracks, "f"));
                }
                disconnectedCallback() {
                    super.disconnectedCallback();
                    __classPrivateFieldGet(this, _instances, "m", _unsub).call(this);
                }
            },
            _tracks = new WeakMap(),
            _unsubscribes = new WeakMap(),
            _subscribe = new WeakMap(),
            _instances = new WeakSet(),
            _sub = function _sub(tracks) {
                __classPrivateFieldGet(this, _unsubscribes, "f").push(...tracks.map(__classPrivateFieldGet(this, _subscribe, "f")));
            },
            _unsub = function _unsub() {
                for (const unsubscribe of __classPrivateFieldGet(this, _unsubscribes, "f"))
                    unsubscribe();
                __classPrivateFieldSet(this, _unsubscribes, [], "f");
            },
            _a;
    };
}
export function mixinSnapstateSubscriptions(...subscriptions) {
    return function (Base) {
        var _unsubscribes_1, _a;
        return _a = class extends Base {
                constructor() {
                    super(...arguments);
                    _unsubscribes_1.set(this, []);
                }
                connectedCallback() {
                    super.connectedCallback();
                    const update = () => { this.requestUpdate(); };
                    __classPrivateFieldSet(this, _unsubscribes_1, subscriptions.map(subscription => subscription(update)), "f");
                }
                disconnectedCallback() {
                    super.disconnectedCallback();
                    for (const unsubscribe of __classPrivateFieldGet(this, _unsubscribes_1, "f"))
                        unsubscribe();
                    __classPrivateFieldSet(this, _unsubscribes_1, [], "f");
                }
            },
            _unsubscribes_1 = new WeakMap(),
            _a;
    };
}
