// xiome/x/features/notes/testing/notes-testing-setup.js
import * as renraku from "renraku";
import { ops } from "../../../framework/ops.js";
import { subbies } from "../../../toolbox/subbies.js";
import { makeNotesModel } from "../models/notes-model.js";
import { mockMeta } from "../../../common/testing/mock-meta.js";
import { makeNotesDepositBox } from "../api/notes-deposit-box.js";
import { makeNotesService } from "../api/services/notes-service.js";
import { prepareMockAuth } from "../../../common/testing/prepare-mock-auth.js";
import { mockStorageTables } from "../../../assembly/backend/tools/mock-storage-tables.js";
import { appPermissions } from "../../../assembly/backend/permissions/standard-permissions.js";
import { UnconstrainedTables } from "../../../framework/api/types/table-namespacing-for-apps.js";
export async function notesTestingSetup() {
    const { rando, appId, config, authPolicies, storage, appOrigin, } = await prepareMockAuth();
    const notesTables = new UnconstrainedTables(await mockStorageTables(storage, {
        notes: true,
        questionDetails: true,
    }));
    const basePolicy = authPolicies.userPolicy;
    const rawNotesService = makeNotesService({
        config,
        notesTables,
        basePolicy,
    });
    const userId = rando.randomId().toString();
    const meta = await mockMeta({
        access: {
            appId: appId.toString(),
            origins: [appOrigin],
            permit: { privileges: [appPermissions.privileges["universal"]] },
            scope: { core: true },
            user: {
                userId: userId.toString(),
                profile: {
                    avatar: undefined,
                    nickname: "Steven Seagal",
                    tagline: "totally ripped and sweet",
                },
                roles: [],
                stats: { joined: Date.now() },
            },
        },
    });
    const headers = { origin: appOrigin };
    const { access } = await basePolicy(meta, headers);
    const notesDepositBox = makeNotesDepositBox({
        rando,
        notesTables: notesTables.namespaceForApp(appId),
    });
    const broadcast = subbies();
    function orchestrateBroadcast(notesModel) {
        notesModel.propagateChangeToOtherTabs.subscribe(() => {
            broadcast.publish(notesModel);
        });
        broadcast.subscribe(originModel => {
            if (notesModel !== originModel)
                notesModel.overwriteStatsOp(originModel.state.statsOp);
        });
    }
    async function browserTab() {
        const notesService = renraku.mock()
            .forService(rawNotesService)
            .withMeta(async () => meta, async () => headers);
        const notesModel = makeNotesModel({ notesService });
        await notesModel.updateAccessOp(ops.ready(access));
        orchestrateBroadcast(notesModel);
        return {
            notesModel,
            makeCacheAlreadySetup() {
                const { cache, setup } = notesModel.createNotesCacheDetails();
                setup();
                return cache;
            },
        };
    }
    return {
        rando,
        access,
        userId,
        backend: { notesDepositBox },
        frontend: await browserTab(),
        browserTab,
    };
}
